src/lib.rs

```diff
@@ -1,718 +1,923 @@
-//! ITM packet parser
+//! A decoder for the ITM and DWT packet protocol as specifed in the
+//! [ARMv7-M architecture reference manual, Appendix
+//! D4](https://developer.arm.com/documentation/ddi0403/ed/). Any
+//! references in this code base refers to this document.
 //!
-//! # References
-//!
-//! - [ARMv7-M Architecture Reference Manual (DDI 0403E.b)][0] - Appendix D4 Debug ITM and DWT
-//! Packet Protocol
-//!
-//! [0]: https://static.docs.arm.com/ddi0403/eb/DDI0403E_B_armv7m_arm.pdf
-//!
-//! - [CoreSight Components Technical Reference Manual (DDI 0314H)][1] - Chapter 12 Instrumentation
-//! Trace Macrocell
-//!
-//! [1]: http://infocenter.arm.com/help/topic/com.arm.doc.ddi0314h/DDI0314H_coresight_components_trm.pdf
+//! *TODO*: fill this out before merge.
 
-#![deny(missing_docs)]
-#![deny(warnings)]
+#[deny(rustdoc::broken_intra_doc_links)]
+mod iter;
+pub use iter::{
+    LocalTimestampOptions, Singles, Timestamp, TimestampConfiguration, TimestampedTracePackets,
+    Timestamps,
+};
 
-use core::fmt;
-use std::io::{self, ErrorKind, Read};
+use std::convert::TryInto;
+use std::io::Read;
+
+use bitmatch::bitmatch;
+use bitvec::prelude::*;
+pub use cortex_m::peripheral::scb::VectActive;
+#[cfg(feature = "serde")]
+use serde_crate::{Deserialize, Serialize};
+
+/// The set of valid packet types that can be decoded.
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum TracePacket {
+    // Synchronization packet category (Appendix D4, p. 782)
+    /// A synchronization packet is a unique pattern in the bitstream.
+    /// It is identified and used to provide the alignment of other
+    /// packet bytes in the bitstream. (Appendix D4.2.1)
+    Sync,
+
+    // Protocol packet category (Appendix D4, p. 782)
+    /// Found in the bitstream if
+    ///
+    /// - Software has written to an ITM stimulus port register when the
+    /// stimulus port output buffer is full.
+    /// - The DWT attempts to generate a hardware source packet when the
+    /// DWT output buffer is full.
+    /// - The local timestamp counter overflows.
+    ///
+    /// See (Appendix D4.2.3).
+    Overflow,
 
-use byteorder::{ByteOrder, LE};
-use either::Either;
-use thiserror::Error;
+    /// A delta timestamp that measures the interval since the
+    /// generation of the last local timestamp and its relation to the
+    /// corresponding ITM/DWT data packets. (Appendix D4.2.4)
+    LocalTimestamp1 {
+        /// Timestamp value.
+        ts: u64, // TODO u32
```

Comment:
The timestamp fits in a `u32`, but some auxilliary functions must be generalized before it can be changed.

Reply:
Fixed; no refactoring required.